### DISCUSSION TITLE
How are release branches supposed to work?

### DISCUSSION THREAD

**Discussion by icj217** [projen/projen#3542] [Q&A] [answered]
I'm a bit confused on how to properly implement the `releaseBranches` feature. My understanding is this:
1. You set the project's major version on the default branch (e.g. main) using `majorVersion`
2. You define separate release branches for historical releases
3. When a PR is opened against the default branch, it'll increment the `majorVersion` line
4. When a PR is opened against a release branch, it'll increment the branch's major version

A couple of questions regarding this:
* Am I correct that release branches are for historical releases lines, rather than forward-looking releases lines?
* At what point in the default branch's history do release branches get created off of? i.e. Do release branches get created BEFORE the major version is bumped on the default branch, or afterwards?
  * My confusion here is specifically related to pull requests, since a PR is going to use the version of the build workflow that exists in the release branch (we currently trigger the release workflow via `workflow_call` from the build workflow to create PR-based pre-releases)

As an example, say a service is going from 1.x => 2.x and we want to continue to contribute changes to the `1.x` release line. This would entail opening a PR that sets the following in projenrc:
```ts
majorVersion: 2,
releaseBranches: {
  '1.x': { majorVersion: 1 }
}
```
Once that's merged, `v2.0.0` is created on the default branch. 

When a new PR is opened against the `1.x` branch, the `release-1.x` workflow doesn't exist on that branch. It only exists on the default branch. According to GHA [docs](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#about-workflow-triggers), this seems problematic since GHA uses the workflow's state at the SHA/branch that triggered the run.

**Comment by mrgrain** [comment]
> Am I correct that release branches are for historical releases lines, rather than forward-looking releases lines?

While that's how they are typically used, release branches also support forward-looking releases lines. Typically you would not set them as default and add a prerelease flag.

> At what point in the default branch's history do release branches get created off of? i.e. Do release branches get created BEFORE the major version is bumped on the default branch, or afterwards?

To be clear, projen doesn't create branches. You will have to create them.

> My confusion here is specifically related to pull requests, since a PR is going to use the version of the build workflow that exists in the release branch (we currently trigger the release workflow via workflow_call from the build workflow to create PR-based pre-releases)

Yes, it's a bit awkward because GitHub Actions will run workflows from two different branches:
- Scheduled workflows, i.e. releases are always run from the workflow on the main branch
- Build workflows are run form the branch

**Comment by mrgrain** [comment, marked as answer]
You should branch off the "old version branch" BEFORE you change the major version:
- `majorVersion: 1` & `defaultReleaseBranch: 'main'` & `releaseBranches: {}`
- Create a new branch `v1`, on that branch change `defaultReleaseBranch: 'v1'`
- Update the default branch: `majorVersion: 2` & `releaseBranches: { '1.x': { majorVersion: 1 } }`

**Reply by icj217** [reply]
Ahh I think the second bullet point is what I was missing. I didn't realize that the dedicated release branch had to be manually modified after the fact, but that makes sense given that the workflow wouldn't be triggered otherwise.

**Reply by mrgrain** [reply]
Yes, and it's super awkward. Which is partly why I recommend the pattern I describe in the next post. To me, it just makes more sense to "move" the current branch in GitHub, than to go back and update an old branch.

**Comment by mrgrain** [comment]
Now my actual recommendation these days is:

- Don't name your default branch `main`, instead name all branches after their version line: `v1`, `v2`, etc.
- In the GitHub Settings, set your default branch to `v1`
- Your `v1` branch will have these settings:
   `majorVersion: 1` & `defaultReleaseBranch: 'v1'` & `releaseBranches: {}`
- When you are preparing the next major version, branch off as `v2`:
  `majorVersion: 2` & `defaultReleaseBranch: 'v2'` & `releaseBranches: { 'v1': { majorVersion: 1 } }`
    (set `prerelease` as needed)
- Now go back to the `v1` branch and add your new branch into `releaseBranches`:
   `majorVersion: 1` & `defaultReleaseBranch: 'v1'` & `releaseBranches: { 'v2': { majorVersion: 2 } }`
- Basically, I recommend to always keep the `releaseBranches` config in sync between all branches!
- Whenever you are ready to actually release your new major version, just change the default branch on GitHub to `v2`
- 🎉 

I use this pattern in my libraries. Have a look here: https://github.com/mrgrain/cdk-esbuild/blob/v5/.projenrc.ts
Just be aware that this is using some custom abstractions to make the branches definition more uniform.
